### PR TITLE
Normalise paths when building BatchFile

### PR DESCRIPTION
**Pull request by frapontillo**
## Problem

When building a `BatchFile` with `LiteBatchFileBuilder`, we were simply checking that `path` wasn't the file separator `/`, so we wouldn't add a duplicated separator after that. But that fix (#475) didn't work for cases where the `path` is something like `my/path/to/the/file/` (with a file separator at the end), so we were essentially still adding a double `/`, causing issues like failed downloads.

## Solution

The solution consists in treating each string we need to concatenate as the same kind of path segment:

1. If the first segment (`storageRoot.path()`) starts with a file separator, we add a file separator to the `StringBuilder`
2. Then, for each element
    1. If the element is empty, we ignore it
    1. We add a file separator to the `StringBuilder` if the last char of the `StringBuilder` is not a file separator
    1. We remove any leading or trailing file separators (`/`) from the segment
    1. We add the segment to the `StringBuilder`

## Tests

Tests have been added to prove that all edge and middle cases are covered for.

### CODE REVIEW

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 1]
```diff
@@ -40,20 +40,60 @@ public BatchFileBuilder saveTo(String path) {
 
     @Override
     public BatchFileBuilder saveTo(String path, String fileName) {
-        this.path = Optional.fromNullable(path).filterNot("/"::equals);
```
Comment (frapontillo) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java:43]: We thought we fixed it by filtering out paths that were `/` only, but this wouldn't solve the issue with a `path` that **ends** in `/` (e.g. `my/path/`).

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    private final InternalBatchBuilder batchBuilder = mock(InternalBatchBuilder.class);
+    private final StorageRoot storageRoot = mock(StorageRoot.class);
+    private final DownloadBatchId downloadBatchId = mock(DownloadBatchId.class);
+
+    private LiteBatchFileBuilder liteBatchFileBuilder;
+
+    @Before
+    public void setUp() {
+        when(downloadBatchId.rawId()).thenReturn("my-movie");
+        liteBatchFileBuilder = new LiteBatchFileBuilder(
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java]: Could move this up to the field and make it `final`.

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    private final InternalBatchBuilder batchBuilder = mock(InternalBatchBuilder.class);
+    private final StorageRoot storageRoot = mock(StorageRoot.class);
+    private final DownloadBatchId downloadBatchId = mock(DownloadBatchId.class);
+
+    private LiteBatchFileBuilder liteBatchFileBuilder;
+
+    @Before
+    public void setUp() {
+        when(downloadBatchId.rawId()).thenReturn("my-movie");
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java]: Can use `BDD` mockito. `given().willReturn`.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 4]
```diff
@@ -40,20 +40,60 @@ public BatchFileBuilder saveTo(String path) {
...
+            if (element.isEmpty()) {
+                continue; // ignore empty paths
+            }
+            if (!isLastCharFileSeparator(stringBuilder)) {
+                stringBuilder.append(File.separatorChar);
+            }
+            stringBuilder.append(removeLeadingTrailingFileSeparator(element));
+        }
+
+        return stringBuilder.toString();
```
Comment (suhtai) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java:76]: To avoid using `StringBuilder` and simplify the logic, we could join all the elements (using something like `TextUtils` or similar) and then remove duplicated separators until none are found and finally any extra wrapping separators.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 5]
```diff
@@ -40,20 +40,60 @@ public BatchFileBuilder saveTo(String path) {
...
+    private String buildPath(String... elements) {
+        StringBuilder stringBuilder = new StringBuilder();
+
+        if (elements[0].startsWith(File.separator)) {
+            stringBuilder.append(File.separator); // if storage path starts with /, we honour it
+        }
+
+        for (String element : elements) {
+            if (element.isEmpty()) {
+                continue; // ignore empty paths
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Could call `elements` `paths`, thereby removing the comment?

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 6]
```diff
@@ -40,20 +40,60 @@ public BatchFileBuilder saveTo(String path) {
...
+        );
 
         parentBuilder.withFile(new BatchFile(networkAddress, downloadFileId, absolutePath));
         return parentBuilder;
     }
 
+    private String buildPath(String... elements) {
+        StringBuilder stringBuilder = new StringBuilder();
+
+        if (elements[0].startsWith(File.separator)) {
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Could pull out the `elements[0]`, name it `storageRootPath` and remove the comment below.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 7]
```diff
@@ -40,20 +40,60 @@ public BatchFileBuilder saveTo(String path) {
...
+            stringBuilder.append(File.separator); // if storage path starts with /, we honour it
+        }
+
+        for (String element : elements) {
+            if (element.isEmpty()) {
+                continue; // ignore empty paths
+            }
+            if (!isLastCharFileSeparator(stringBuilder)) {
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Could invert the statement inside this method and rename to remove the `!` negation

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 8]
```diff
@@ -40,20 +40,60 @@ public BatchFileBuilder saveTo(String path) {
...
+        int beginIndex = 0;
+        int endIndex = element.length();
+
+        if (element.charAt(0) == File.separatorChar) {
+            beginIndex = 1;
+        }
+        int lastIndexOfSeparator = element.lastIndexOf(File.separatorChar);
+        if (lastIndexOfSeparator != 0 && lastIndexOfSeparator == element.length() - 1) {
+            endIndex = element.length() - 1;
+        }
```
Comment (suhtai) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java:91]: This check could just check the last character of the element:
```
if(element.charAt(endIndex - 1) == File.separatorChar) {
  endIndex++
}
```
Both start and end could also use a `while` to check multiple ones:
```
while(element.charAt(beginIndex) == File.separatorChar && beginIndex < endIndex) {
  beginIndex++;
}
while(element.charAt(endIndex - 1) == File.separatorChar && beginIndex < endIndex) {
  endIndex--;
}
```

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 9]
```diff
@@ -40,20 +40,60 @@ public BatchFileBuilder saveTo(String path) {
...
+                downloadBatchId.rawId(),
+                path.or(""),
+                fileName.getOrElse(() -> FileNameExtractor.extractFrom(networkAddress))
+        );
 
         parentBuilder.withFile(new BatchFile(networkAddress, downloadFileId, absolutePath));
         return parentBuilder;
     }
 
+    private String buildPath(String... elements) {
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: The only problem with `String... elements` is that you can change the order and everything breaks 😄 but I guess this is backed with tests so 🤷‍♂️ 😄

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java (hunk 2 above)
Comment (suhtai) [reply library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java]: Or move it locally with a `given` factory method which takes in the conditions of the subject

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
+package com.novoda.downloadmanager;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class LiteBatchFileBuilderTest {
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java:10]: Wondering if we should have a test / tests with a ludicrous number of Separators spaced throughout the `root` and the `paths`? At the moment the most extreme case we have is `/path/`.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 4 above)
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java:76]: I like the idea, it might make it a bit simpler. I wonder how heavy `TextUtils` or similar is compared to this method 🤔

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 11]
```diff
@@ -40,20 +40,60 @@ public BatchFileBuilder saveTo(String path) {
...
+            beginIndex = 1;
+        }
+        int lastIndexOfSeparator = element.lastIndexOf(File.separatorChar);
+        if (lastIndexOfSeparator != 0 && lastIndexOfSeparator == element.length() - 1) {
+            endIndex = element.length() - 1;
+        }
+        return element.substring(beginIndex, endIndex);
+    }
+
+    private boolean isLastCharFileSeparator(StringBuilder stringBuilder) {
```
Comment (suhtai) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: This method can take a `CharSequence` Always use the most abstract form of a type possible (Effective Java)

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 5 above)
Comment (suhtai) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: `continue` == `goto` :trollface:

**Review by MechaRyRy — COMMENTED** [review]
This is definitely heading in the right direction 💯 ❤️

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 4 above)
Comment (frapontillo) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java:76]: I personally find the current approach easier to understand, as we are building the string one step at a time, instead of doing backtracking on something already built.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 5 above)
Comment (frapontillo) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: 😄 I would have used `filter` but you know, Java.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 8 above)
Comment (frapontillo) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java:91]: In the first snippet, you probably meant `endIndex--`. Anyway, it would crash for the case of element being `/`, since `beginIndex` would be `1` and `endIndex` would end up being `0`.

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java (hunk 10 above)
Comment (frapontillo) [reply library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java]: So, I agree we could do that (sanitise all elements throughout) but I think that could also be the user's job. I think accommodating for leading and trailing separators is enough (for now) to fix some weird edge cases.

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java (hunk 2 above)
Comment (frapontillo) [reply library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java]: Done in 469aca79cb5739c28aa31d4338a2022f46d89ceb.

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java (hunk 3 above)
Comment (frapontillo) [reply library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java]: Done in d5a16cbdfb291c625e4d91fccd0c0d3d7bc269fc.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 5 above)
Comment (frapontillo) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Done in 9c2bd121b95df17de9e0497cf177497b106bf482.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 7 above)
Comment (frapontillo) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Done in 8998736eea676b0be82a8a1d7c4d4bed9eeff0ff.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 11 above)
Comment (frapontillo) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Done in 8c627c326481c8c56968bc71bab1ea49b4d5da92.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 6 above)
Comment (frapontillo) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Done in 3dca3ed8c1cca7699221c07a85aae2ed99525203.

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java (hunk 10 above)
Comment (frapontillo) [reply library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java:10]: Hmmm, I have slightly changed `doesNotAddDuplicateSeparators` in bd9a16a824a8926f8bb08bb9521220996417fe05 to test the case where every single path element has leading and trailing separators.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 8 above)
Comment (suhtai) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java:91]: `beginIndex <= endIndex` then? 😁

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 12]
```diff
@@ -40,20 +40,61 @@ public BatchFileBuilder saveTo(String path) {
...
+        StringBuilder stringBuilder = new StringBuilder();
+
+        String storagePath = paths[0];
+        if (storagePath.startsWith(File.separator)) {
+            stringBuilder.append(File.separator);
+        }
+
+        for (String path : paths) {
+            if (path.isEmpty()) {
+                continue; // ignore empty paths
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Didn't remove the comment 👅

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java (hunk 10 above)
Comment (MechaRyRy) [reply library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java:10]: Sorry, my comment was hinting at what happens if you have `//path//` for example.

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+        verify(batchBuilder).withFile(batchFileWithPath("root/my-movie/my/path/my-movie.mp4"));
+    }
+
+    @Test
+    public void doesNotAddDuplicateSeparators() {
+        given(storageRoot.path()).willReturn("/root/");
+        given(downloadBatchId.rawId()).willReturn("/my-movie/");
+
+        liteBatchFileBuilder
```
Comment (suhtai) [inline library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java]: I meant to have the builder as a local property in the test functions 😁

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java (hunk 12 above)
Comment (frapontillo) [reply library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: Oh, i thought the comment was for the `storagePath`!

**Comment by frapontillo** [comment]
Implementation changed in 0224efaef89fd01b8e8c951efaeff7cf3cf4b5da to clean up internal duplicated separators too 💯.

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 14]
```diff
@@ -40,20 +42,62 @@ public BatchFileBuilder saveTo(String path) {
...
 
+    private String buildPath(String... paths) {
+        StringBuilder stringBuilder = new StringBuilder();
+
+        String storagePath = paths[0];
+        if (storagePath.startsWith(File.separator)) {
+            stringBuilder.append(File.separator);
+        }
+
+        String joinedPaths = sanitise(joinWithFileSeparator(paths));
```
Comment (frapontillo) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java:73]: The new logic ✨

File: library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java [hunk 15]
```diff
@@ -40,20 +42,62 @@ public BatchFileBuilder saveTo(String path) {
...
+        List<CharSequence> filteredPathSegments = new ArrayList<>();
+        for (String pathSegment : pathSegments) {
+            if (!pathSegment.isEmpty()) {
+                filteredPathSegments.add(pathSegment);
+            }
+        }
+        return filteredPathSegments.toArray(new CharSequence[0]);
+    }
+
+    private String joinWithFileSeparator(CharSequence[] elements) {
```
Comment (frapontillo) [inline library/src/main/java/com/novoda/downloadmanager/LiteBatchFileBuilder.java]: I opted to re-implement the `join` method since the `TextUtils` one can't be used in tests and `String.join` is definitely not available for our min API (16).

File: library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        given(storageRoot.path()).willReturn("/root/");
+        given(downloadBatchId.rawId()).willReturn("/my-movie/");
+
+        whenLiteBatchFileBuilderApply("/my/path/", "/my-movie.mp4");
+
+        verify(batchBuilder).withFile(batchFileWithPath("/root/my-movie/my/path/my-movie.mp4"));
+    }
+
+    @Test
+    public void doesNotAddDuplicateSeparatorsWhenInMiddleOfSegments() {
```
Comment (frapontillo) [inline library/src/test/java/com/novoda/downloadmanager/LiteBatchFileBuilderTest.java:48]: Here's your ludicrous test @Mecharyry 😆

**Comment by frapontillo** [comment]
We leave the developer to specify the file path as best as possible. This would only provide some safety around what's been passed, which should be separated with a `File.separator`.

**Review by MechaRyRy — DISMISSED** [review]
❤️ great stuff ❤️

**Review by suhtai — APPROVED** [review]

**Review by MechaRyRy — APPROVED** [review]
